### ISSUE TITLE
Http request to Https Server produces 7-byte response

### ISSUE THREAD

**Issue by colinsmcintyre** [eclipse/jetty.project#1095]
Hello,

We use embedded Jetty in a Springboot application and have encountered an odd issue.  We have SSL configured, however when an HTTP request is made Jetty responds with a 7-byte response rather than an empty response.  This in turn causes the browser/client to react to the 7 bytes rather than providing an ERR_EMPTY_RESPONSE as expected.  In Chrome this results in a the 7 bytes downloading; in IE the bytes are shown as much as possible; using curl the bytes are returned.

This began for us about a month ago, when we upgraded our version of Springboot (to patch a Jackson vulnerability), which took us from Jetty9.2.19.v20160908 to Jetty9.3.11.v20160721.  On further investigation Jetty9.3.9.v20160517 does not present the problem, but Jetty9.3.10.v20160621 does.

**Comment by colinsmcintyre** [comment]
[7Bytes.txt](https://github.com/eclipse/jetty.project/files/589160/7Bytes.txt)
![7bytes](https://cloud.githubusercontent.com/assets/9606531/20264468/9df54c0c-aa63-11e6-900c-e55c6f0028e5.PNG)

Text file with the 7Bytes, and image of what they look like in Notepad++

**Comment by colinsmcintyre** [comment]
[Jetty9.3_SSLTest.zip](https://github.com/eclipse/jetty.project/files/589162/Jetty9.3_SSLTest.zip)

Example Maven project to demonstrate the issue.  Run the project (i used NetBeans) and compare http://localhost:8080/about with https://localhost:8080/about

**Comment by sbordet** [comment]
The bytes are:

```
15 03 03 00 02 02 50
```

where:

`15` == TLS Alert Message Type
`03 03` == TLS Protocol version 1.2
`00 02` == TLS Frame length
`02` == Alert level Fatal
`50` == Internal Error

**Comment by sbordet** [comment]
Are you connecting using the `http` scheme to the TLS port ?

**Comment by joakime** [comment]
Seems like the behavior with accessing a TLS port with a non-TLS scheme is undefined.

```
$ curl http://webtide.com:443/


$ curl http://google.com:443/
curl: (52) Empty reply from server

$ curl http://facebook.com:443/
curl: (56) Recv failure: Connection reset by peer

$ curl http://salesforce.com:443/
(

$ curl http://wikipedia.org:443/
<html>
<head><title>400 The plain HTTP request was sent to HTTPS port</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<center>The plain HTTP request was sent to HTTPS port</center>
<hr><center>nginx/1.11.6</center>
</body>
</html>

$ curl http://bbc.co.uk:443/
curl: (52) Empty reply from server

$ curl http://youtube.com:443/
curl: (52) Empty reply from server

$ curl http://ietf.org:443/
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<title>Bad request!</title>
<link rev="made" href="mailto:ietf-action@ietf.org" />
<style type="text/css"><!--/*--><![CDATA[/*><!--*/ 
    body { color: #000000; background-color: #FFFFFF; }
    a:link { color: #0000CC; }
    p, address {margin-left: 3em;}
    span {font-size: smaller;}
/*]]>*/--></style>
</head>

<body>
<h1>Bad request!</h1>
<p>


    Your browser (or proxy) sent a request that
    this server could not understand.

</p>
<p>
If you think this is a server error, please contact
the <a href="mailto:ietf-action@ietf.org">webmaster</a>.

</p>

<h2>Error 400</h2>
<address>
  <a href="/">www6.ietf.org</a><br />
  <span>Apache</span>
</address>
</body>
</html>

$ curl http://microsoft.com:443/


$ curl http://mozilla.com:443/
curl: (52) Empty reply from server

$ curl http://eclipse.org:443/
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>400 Bad Request</title>
</head><body>
<h1>Bad Request</h1>
<p>Your browser sent a request that this server could not understand.<br />
Reason: You're speaking plain HTTP to an SSL-enabled server port.<br />
Instead use the HTTPS scheme to access this URL, please.<br />
<blockquote>Hint: <a href="https://www.eclipse.org/"><b>https://www.eclipse.org/</b></a></blockquote></p>
</body></html>
```

**Comment by colinsmcintyre** [comment]
> Are you connecting using the http scheme to the TLS port ?

Yes.  Our concern is that if a user does this by accident, there's nothing clear to tell the user they've done the wrong thing.

**Comment by sbordet** [comment]
I don't understand the concern.

If your user follows links or such, then it's the application that must provide the right links.

If your user types wrong stuff in the browser address bar, then he/she may even mistype the domain and well, end up elsewhere.

**Comment by colinsmcintyre** [comment]
Yes, the user can always type in an incorrect address but in the majority of cases it's obvious that the user is in the wrong, be it an empty response or the wrong page displayed.  In this case I don't think it is obvious to the user (Chrome downloads a file and IE displays a white page with a "P").  My immediate reaction was "what's wrong with the server/application?".

**Comment by laeubi** [comment]
The other way round (https on http connector) it seems jetty already detects this and issues a warning in the server-log.

**Comment by gregw** [comment]
It is interesting that some servers at least are able to respond with a valid HTTP response describing the error.

It is conceivable that we could do the same, however with the advent of  ALPN and HTTP/2 it is not really clear that sending a HTTP with HTML response will always be appropriate.

The issue is that as we negotiate the TLS connection, we do not yet know what protocol (HTTP/1, SPDY, HTTP/2, something new?) that we are being asked to speak.   If the parsing of the TLS Client Hello fails, then I guess we could examine the received bytes and look for a 'G' 'E' 'T' and send a HTTP error.

But equally the browsers could look at a non HTTP response and check for a TLS message alert and display a warning.

So this is something that would be nice to have, but as this code is already complex and has to soon cope with JDK 9 changes, I'm not sure it is exactly a high priority for us to implement.

Maybe once JDK 9 ALPN changes have landed, we can include better error handling in the reworking we have to do for that?

**Comment by joakime** [comment]
@gregw @sbordet now that JDK9 (and JDK10 is due out in few weeks) is this something we should consider adding to our alpn layer for JDK9?

I can see this being just a small file (static content, HTTP headers + content, then close) being returned on the connection and then closing.
(mimicing the wikipedia approach)

**Comment by sbordet** [comment]
I don't think ALPN is involved here.
The whole thing should reside in `SslConnection` where we try the first unwrap, which will fail with an exception, and at that point we examine the network buffer.
If the network buffer contains `'G' 'E' 'T' ' '`, then we shove a small clear-text response and we close.

**Comment by joakime** [comment]
How about if we cannot unwrap, we just process the 400 response via our normal 400 response techniques, but in clear text?

**Comment by sbordet** [comment]
We certainly fail the `unwrap()` and yes we shove a 400 response in clear text in `SslConnection._encryptedOutput`.
